x/examples/outline-connectivity-app/shared_backend/main.go

```diff
@@ -0,0 +1,232 @@
+// Copyright 2023 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shared_backend
+
+import (
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"log"
+	"net"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/transport"
+	"github.com/Jigsaw-Code/outline-sdk/transport/shadowsocks"
+	"github.com/Jigsaw-Code/outline-sdk/x/config"
+	"github.com/Jigsaw-Code/outline-sdk/x/connectivity"
+
+	_ "golang.org/x/mobile/bind"
+)
+
+type ConnectivityTestProtocolConfig struct {
+	TCP bool `json:"tcp"`
+	UDP bool `json:"udp"`
+}
+
+type ConnectivityTestResult struct {
+	// Inputs
+	Proxy    string `json:"proxy"`
+	Resolver string `json:"resolver"`
+	Proto    string `json:"proto"`
+	Prefix   string `json:"prefix"`
+	// Observations
+	Time       time.Time              `json:"time"`
+	DurationMs int64                  `json:"durationMs"`
+	Error      *ConnectivityTestError `json:"error"`
+}
+
+type ConnectivityTestError struct {
+	// TODO: add Shadowsocks/Transport error
+	Op string `json:"operation"`
+	// Posix error, when available
+	PosixError string `json:"posixError"`
+	// TODO: remove IP addresses
+	Msg string `json:"message"`
+}
+
+type ConnectivityTestRequest struct {
+	AccessKey string                         `json:"accessKey"`
+	Domain    string                         `json:"domain"`
+	Resolvers []string                       `json:"resolvers"`
+	Protocols ConnectivityTestProtocolConfig `json:"protocols"`
+}
+
+type sessionConfig struct {
+	Hostname  string
+	Port      int
+	CryptoKey *shadowsocks.EncryptionKey
+	Prefix    Prefix
+}
+
+type Prefix []byte
+
+func ConnectivityTest(request ConnectivityTestRequest) ([]ConnectivityTestResult, error) {
+	accessKeyParameters, err := parseAccessKey(request.AccessKey)
```

Comment:
This code is tied to Shadowsocks. I'd like to see this use the `x/config` package instead, so it can showcase the different transports. See https://github.com/Jigsaw-Code/outline-sdk/blob/main/x/examples/outline-connectivity/main.go.

But let's do that in a different PR.